logictest: randomly backup/restore in logic tests

pbardea opened the pull request:
This commit introduces a new configuration to logic tests that will
randomly backup the cluster, re-create the cluster and restore that
backup to the new cluster.

This is primarily beneficial since it runs backup/restore under a much
more comprehensive set of SQL states.

Closes #54060.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54893)
<!-- Reviewable:end -->

pbardea committed:
logictest: randomly backup/restore in logic tests

This commit introduces a new configuration to logic tests that will
randomly backup the cluster, re-create the cluster and restore that
backup to the new cluster.

This is primarily beneficial since it runs backup/restore under a much
more comprehensive set of SQL states.

Closes #54060.

Release note: None

pbardea committed:
logictest: seperate test and cluster setup

This commit seperates the creation/deletion of a cluster from the setup
of a logicTest. This work is to enable backup/restore to be integrated
in logic tests. The plan is to randomly backup and restore the cluster
between statements in the logic tests, however cluster restore requires
a "fresh" cluster. This work enables future work to reset a cluster
within a logic test.

Release note: None

pbardea committed:
sql: skip backup-restore logic test profile on incompatible tests

This commit is an initial conservative skipping of many logic tests that
don't pass under the backup-restore logic test configuration. These
tests typically don't pass that configuration because of:
- They inspect the jobs table which changes during a backup/restorer
- They inspect descriptor IDs directly. Descriptors will have different
IDs when created after a backup/restore will be different than if no
backup/restore was taken (due to the temporary tables created during the
restore).
- They use prepared statements which are not preserved during a restore.

Release note: None

pbardea commented:
The failing CI needs a bit of investigation, but I think that this is ready for an initial look. 

Also, for the logic test files that fail I will create an issue outlining why they fail and link them in the !3node-backup logic test directive.

I'll also squash the 2 commits so that the logic tests pass on every commit, but I think keeping the commits separate makes it easier to review.

pbardea commented:
Ooops, forgot to add the rest of the team. Curious if others have thoughts on the first commit. I'm planning on looking into making CI happy this aft.

pbardea commented:
Broke out the first commit into #54893.
